state-chain/pallets/cf-elections/src/electoral_systems/solana_swap_accounts_tracking.rs

```diff
@@ -0,0 +1,214 @@
+};
+use cf_chains::sol::{
+	api::ContractSwapAccountAndSender, MAX_BATCH_SIZE_OF_CONTRACT_SWAP_ACCOUNT_CLOSURES,
+	MAX_WAIT_BLOCKS_FOR_SWAP_ACCOUNT_CLOSURE_APICALLS,
+	NONCE_AVAILABILITY_THRESHOLD_FOR_INITIATING_SWAP_ACCOUNT_CLOSURES,
+};
+use cf_utilities::success_threshold_from_share_count;
+use frame_support::{
+	pallet_prelude::{MaybeSerializeDeserialize, Member},
+	sp_runtime::traits::CheckedSub,
+	Parameter,
+};
+use itertools::Itertools;
+use sp_std::vec::Vec;
+
+pub trait SolanaVaultSwapAccountsHook<Account, SwapDetails, E> {
+	fn close_accounts(accounts: Vec<Account>) -> Result<(), E>;
+	fn initiate_vault_swap(swap_details: SwapDetails);
+	fn get_number_of_available_sol_nonce_accounts() -> usize;
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize, TypeInfo, Encode, Decode)]
+pub struct SolanaVaultSwapsElectoralState<Account: Ord, BlockNumber> {
+	pub block_number_last_closed_accounts: BlockNumber,
+	pub witnessed_open_accounts: Vec<Account>,
+	pub closure_initiated_accounts: BTreeSet<Account>,
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+impl BenchmarkValue for SolanaVaultSwapsElectoralState<ContractSwapAccountAndSender, u32> {
+	fn benchmark_value() -> Self {
+		Self {
+			block_number_last_closed_accounts: 1u32,
+			witnessed_open_accounts: vec![BenchmarkValue::benchmark_value()],
+			closure_initiated_accounts: BTreeSet::from([BenchmarkValue::benchmark_value()]),
+		}
+	}
+}
+
+#[derive(
+	Clone, PartialEq, Eq, Debug, Serialize, Deserialize, TypeInfo, Encode, Decode, Ord, PartialOrd,
+)]
+pub struct SolanaVaultSwapsVote<Account: Ord, SwapDetails: Ord> {
+	pub new_accounts: BTreeSet<(Account, SwapDetails)>,
+	pub confirm_closed_accounts: BTreeSet<Account>,
+}
+
+pub struct SolanaVaultSwapAccounts<
+	Account,
+	SwapDetails,
+	BlockNumber,
+	Settings,
+	Hook,
+	ValidatorId,
+	E,
+> {
+	_phantom: core::marker::PhantomData<(
+		Account,
+		SwapDetails,
+		BlockNumber,
+		Settings,
+		Hook,
+		ValidatorId,
+		E,
+	)>,
+}
+impl<
+		E: sp_std::fmt::Debug + 'static,
+		Account: MaybeSerializeDeserialize + Member + Parameter + Ord,
+		SwapDetails: MaybeSerializeDeserialize + Member + Parameter + Ord,
+		BlockNumber: MaybeSerializeDeserialize + Member + Parameter + Ord + CheckedSub + Into<u32> + Copy,
+		Settings: Member + Parameter + MaybeSerializeDeserialize + Eq,
+		Hook: SolanaVaultSwapAccountsHook<Account, SwapDetails, E> + 'static,
+		ValidatorId: Member + Parameter + Ord + MaybeSerializeDeserialize,
+	> ElectoralSystem
+	for SolanaVaultSwapAccounts<Account, SwapDetails, BlockNumber, Settings, Hook, ValidatorId, E>
+{
+	type ValidatorId = ValidatorId;
+	type ElectoralUnsynchronisedState = SolanaVaultSwapsElectoralState<Account, BlockNumber>;
+	type ElectoralUnsynchronisedStateMapKey = ();
+	type ElectoralUnsynchronisedStateMapValue = ();
+
+	type ElectoralUnsynchronisedSettings = ();
+	type ElectoralSettings = Settings;
+	type ElectionIdentifierExtra = ();
+	type ElectionProperties = ();
```

Comment:
I'm surprised there are no election properties. 

I'm guessing the intention is to have the engine read the `ElectoralUnsynchronisedState` but this is not how the electoral systems are intended to work. 

All the required context for the vote (i'm guessing this is: 'currently opened accounts') should be stored in the properties: see the fn signature of the voting api.

Reply:
I considered using Properties since we have access to it in the Voting api but didnt want to store the list twice (once in the unsynchronised state and once in properties). But I can see now that if we use the properties, we dont need to use the unsynchronised state.

Reply:
You can think of Properties as a way to pass arguments to the engine voting rather than storage.

Reply:
That's true, but to expand on *why* the properties are there and separate: The issue is in the name "unsynchronised" - the engines need to receive the same election inputs to provide consistent witness data. By using the properties it is then synchronised by the election identifier.

Reply:
good to know 👍 However, in this case the unsynchronised state is also synchronised since there is only one election at a given moment and the only way it changes is when a consensus is reached and we then immediately delete that election and start a new one. 

But in any case, we can use properties here and then we dont need to use the unsynchronised state. will do that.